pkg/addons/device_plugin.go

```diff
@@ -197,11 +209,59 @@ func (n *NvidiaDevicePlugin) Deploy() error {
 	return applyDevicePlugin(n)
 }
 
+// SetTolerations sets given tolerations on the DaemonSet if they don't already exist.
+// We check the taints on each node which is an NVIDIA instance type and apply
+// tolerations for all the taints defined on the node.
+func (n *NvidiaDevicePlugin) SetTolerations(spec *corev1.PodTemplateSpec) error {
+	contains := func(list []corev1.Toleration, key string) bool {
+		for _, t := range list {
+			if t.Key == key {
+				return true
+			}
+		}
+		return false
+	}
+	// don't duplicate taints from other nodes or overwrite them with
+	// different values ( shouldn't happen in general... )
+	taints := make(map[string]api.NodeGroupTaint)
+	for _, ng := range n.spec.NodeGroups {
+		if api.HasInstanceType(ng, instance.IsNvidiaInstanceType) &&
+			ng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2 {
+			for _, taint := range ng.Taints {
+				if _, ok := taints[taint.Key]; !ok {
+					taints[taint.Key] = taint
+				}
+			}
+		}
+	}
+	for _, ng := range n.spec.ManagedNodeGroups {
+		if api.HasInstanceTypeManaged(ng, instance.IsNvidiaInstanceType) &&
+			ng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2 {
+			for _, taint := range ng.Taints {
+				if _, ok := taints[taint.Key]; !ok {
+					taints[taint.Key] = taint
+				}
+			}
+		}
+	}
+	for _, t := range taints {
+		// only add toleration if it doesn't already exist. In that case, we don't overwrite it.
+		if !contains(spec.Spec.Tolerations, t.Key) {
+			spec.Spec.Tolerations = append(spec.Spec.Tolerations, corev1.Toleration{
+				Key:   t.Key,
+				Value: t.Value,
+			})
+		}
+	}
+	return nil
+}
+
 // A EFADevicePlugin deploys the EFA Device Plugin to a cluster
 type EFADevicePlugin struct {
 	rawClient kubernetes.RawClientInterface
 	region    string
 	planMode  bool
+	spec      *api.ClusterConfig
```

Comment:
We don't need to add this field since it isn't used.

Reply:
So I was thinking about this. I only added it because of consistency, but do you think we should actually do this for other plugins as well? Won't they eventually face the same issue?

Reply:
Right, other plugins will also face the same issue, but it might be out of scope for this PR. If we're adding this to other plugins, I'd say we either add the full implementation or leave them as-is.

Reply:
True that.